elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/dimension/AbstractDimension.java

```diff
@@ -0,0 +1,110 @@
+        List<String> singleFriendlyName = entityDictionary.getAllFields(cls).stream()
+                .filter(field -> entityDictionary.attributeOrRelationAnnotationExists(cls, field, FriendlyName.class))
+                .collect(Collectors.toList());
+
+        if (singleFriendlyName.size() > 1) {
+            String message = String.format(
+                    "Multiple @FriendlyName fields found in entity '%s'. Can only have 0 or 1",
+                    cls.getName()
+            );
+            log.error(message);
+            throw new IllegalStateException(message);
+        }
+
+        return singleFriendlyName.size() == 1 ? singleFriendlyName.get(0) : null;
+    }
+
+    /**
+     * Returns the specified cardinality of this {@link Dimension}, {@link #getDefaultCardinality() default} is returned
+     * when the dimension is not annotated by {@link Cardinality}.
+     *
+     * @param cls  The entity name of the dimension from which estimated cardinality is being looked up
+     * @param entityDictionary  The auxiliary object that offers binding info about the cardinality for this dimension
+     *
+     * @return dimension size
+     */
+    public static DimensionSize getEstimatedCardinality(Class<?> cls, EntityDictionary entityDictionary) {
+        Cardinality annotation = entityDictionary.getAnnotation(cls, Cardinality.class);
+        return annotation == null ? getDefaultCardinality() : annotation.size();
+    }
+
+    /**
+     * The default cardinality for entity without {@link Cardinality} annotation.
+     *
+     * @return the default table size backing this {@link Dimension}
+     */
+    public static DimensionSize getDefaultCardinality() {
+        return DimensionSize.LARGE;
+    }
+
+    private static final long serialVersionUID = -3072568656514365362L;
+
+    @Getter
+    private final String dimensionIdentifier;
+
+    private final String friendlyName;
+
+    @Getter
+    private final DimensionSize cardinality;
+
+    /**
+     * Constructor.
+     *
+     * @param dimensionIdentifier  The relationship/entity name of this dimension
+     * @param friendlyName  The name of the field, whose value represents a human-readable name of of this dimension
+     * @param cardinality  The estimated size of the DB table backing this dimension
+     *
+     * @throws NullPointerException if {@code dimensionIdentifier} or {@code cardinality} is {@code null}
+     */
+    public AbstractDimension(
+            final String dimensionIdentifier,
+            final String friendlyName,
+            final DimensionSize cardinality
+    ) {
+        this.dimensionIdentifier = Objects.requireNonNull(dimensionIdentifier, "dimensionIdentifier");
+        this.friendlyName = friendlyName;
+        this.cardinality = Objects.requireNonNull(cardinality, "cardinality");
+    }
+
+    @Override
+    public Optional<String> getFriendlyName() {
+        return friendlyName == null ? Optional.empty() : Optional.of(friendlyName);
```

Comment:
There should always be a friendly name.  It will fall into one of three categories:

1. A dimension table has a single attribute with the friendly name annotation.  That attribute is the friendly name.
2. A dimension table has no attribute with the friendly name annotation.  We should use its `@Id` field as the friendly name.
3. A fact table has a degenerate dimension (an attribute which is also a dimension).  The friendly name is the attribute in that case.  There should not be an annotation here.

Reply:
Thanks